Language updates (#6)
apexlang/apexlang.io: docs/specification.md

By default, a declared type is required and if unset, contains a zero value (`0`, `""`, empty array or map). To make any type optional (nullable), follow it with a `?`. For example,  `string?` represents an optional string value.

### Functions

Functions are independent operations and allow two patterns for parameter input.

> Common functions accept zero-to-many parameters and are generated with the same signature defined in Apex. Functions are recommended when passing in a small number of fields and follow a simple, familiar, and easy to read format. All parameters are named. In this example, we pass in first and last names to create a customer and return its new identifier as a `u64`.

Reviewer: ```suggestion
Common functions accept zero or more parameters and are generated with the same signature defined in Apex. Functions are recommended when passing in a small number of fields and follow a simple, familiar, and easy to read format. All parameters must be named. In this example, we pass in first and last names to create a customer and return its new identifier as a `u64`.
```

```apexlang
func createCustomer(firstName: string, lastName: string): u64
```

> Now, let's say you want to model a request/reply style operation or you want to accept a large number of fields inside a wrapper object. This is where [unary operations](https://en.wikipedia.org/wiki/Unary_operation) are preferred. In contrast to functions, unary operations accept exactly one input.

Reviewer: would there ever be a need to pass two wrapper objects?

example: `Customer` and `Item`?

Author: In that case you would either use a common operation with 2 params (recommended) or create a wrapper for the wrappers.

Instead of using parenthesis `(...)` to enclose parameters, unary operations use brackets `[...]` to enclose a single parameter. This is what `createCustomer` would look like accepting a Customer data type. We'll look at how `Customer` is defined shortly.

```apexlang
func createCustomer[customer: Customer]: u64
```

> Why the different syntax? In the case of common functions, multiple parameters are possible so the arguments are likely encapsulated in a wrapper object for serialization. For unary operations, since there is a single parameter no wrapper object is required and the input object can be serialized directly. This syntax is necessary for the the code generation tools to generate the appropriate client and server code.

Reviewer: This is not clear to me.

- can I only have a unary operation if I used a custom (non-default) data type?  Or can I have a unary with a `string`?
- why can the codegen not look to see that a function has exactly 1 parameter (that is not a default type) and treat that as unary?

Author: > can I only have a unary operation if I used a custom (non-default) data type? Or can I have a unary with a string?

Any type is allowing, including streams

> why can the codegen not look to see that a function has exactly 1 parameter (that is not a default type) and treat that as unary?

In that scenario, if a second argument with a default value is added then a wrapper is produced and creates a breaking change for existing clients.

### Interfaces

> Interfaces are conceptual groups of operations that allow the developer divide communication up into any number of components. Typically, interfaces are named according to their purpose. For example, a distributed calculator could split each mathematical operation up into different roles implemented by different modules.

Reviewer: ```suggestion
Interfaces are conceptual groups of operations that allow the developer to divide communication into multiple components. Typically, interfaces are named according to their purpose. For example, a distributed calculator could split each mathematical operation into different roles implemented by different modules.
```

Reviewer: What is a `role` and what is a `module`?  Are these generic terms or is there any definition that is significant specifically to apex lang?

}
```

### Default values

> Fields can also specify default values when a type is initialized. This needs to be carefully considered in the code generation. Languages vary of how this would be implemented.

Reviewer: ```suggestion
Default values for Fields can specified when a type is initialized.  Codegen implementation for default values may vary based on limitations of the target language.
```

```apexlang
type PhoneNumber {
  number: string
  type:   PhoneType
}

enum PhoneType {
  mobile = 0 as "Mobile"
  home   = 1 as "Home"
  work   = 2 as "Work"
}
```

Each enum value denotes its programatic / variable name, the integer value that is serialized, and a display or friendly name for printing. Note that Apex does not address internationalization so custom code is required to print the value in multiple spoken languages.

### Union types

Unions types denote that a type can have one of several representations.

```apexlang
union Animal = Cat | Dog
```

The code generator must provide a way to determine the actual type at runtime and handle serialization if necessary.

### Alias types

> Alias types are used for cases when scalar types (like string) should be parsed our treated like a different data type in the generated code. For example, [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier)s are commonly serialized as strings, but parsed into a specific data type.

Reviewer: ```suggestion
Alias types are used for cases when scalar types (like string) should be parsed or treated like a different data type in the generated code. For example, [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier)s are commonly serialized as strings, but parsed into a specific data type.
```

The annotation examples above shows a validation scenario but annotations are not limited to this purpose. The developer has the freedom to extend the code generation tool to leverage annotations for their application's needs. In the `Customer` example, the developer could use these annotations to generate `validate` methods on each of the generated object types.

> Directives are used to ensure that an annotation's arguments match and an expected format. For constraining a numeric or length range, the directive for `@range` might look like this:

Reviewer: ```suggestion
Directives are used to ensure that an annotation's arguments match an expected format. For constraining a numeric or length range, the directive for `@range` might look like this:
```

```apexlang
directive @range(min: u32, max: u32)
  on FIELD
  require @valid on TYPE
```

> This directive specifies parameters but also where it can be applied. The syntax `on FIELD require @valid on TYPE` says "only allow this on fields where there is also an @valid annotation on its type". Multiple locations are supported and delimited with `|`:

Reviewer: ```suggestion
This directive specifies parameters but also where it can be applied. The syntax `on FIELD require @valid on TYPE` means "only allow this on fields where there is also an @valid annotation on its type". Multiple locations are supported and delimited with `|`:
```

Reviewer: This section seems to be lacking some details.  I am not sure what `@valid` means.

Can you make this refer to a previously used example?  Also, this doesn't make sense to me.  Can you put a `@valid` annotation on a `TYPE`?  How do would you use this on `NAMESPACE` or `PARAMETER`?

```apexlang
NAMESPACE | INTERFACE | OPERATION | PARAMETER | TYPE | FIELD | ENUM | ENUM_VALUE | UNION | ALIAS
```

> Directive improve the user experience of annotations by giving the developer useful error information when thier annotations are invalid. These directives might be something made available through imports.

Reviewer: ```suggestion
Directives improve the user experience of annotations by giving the developer useful error details when the annotations are invalid.  Directives can be in the main file or be made available through imports.
```

### Imports

> Apex makes use of installable modules that can contain useful definitions to import. For example, OpenAPI annotations are imported with the following statement:

Reviewer: ```suggestion
Apex can integrate external modules using the `import` command. For example, OpenAPI annotations can be imported with the following statement:
```

```apexlang
import * from "@apexlang/openapi"
```

> You are also able to import specific definitions:

Reviewer: ```suggestion
Is is also possible to import specific definitions:
```

> ```apexlang

Reviewer: add example of importing directives as mentioned above?

Author: In the example here `@info` is a directive. I could add the example using it on a `namespace`.

```apexlang
type Customer {
  firstName:  string        @notEmpty
  middleName: string?       @notEmpty
  lastName:   string        @notEmpty
  address1:   string        @notEmpty
  address2:   string?       @notEmpty
  city:       string        @length(2)
  zipcode:    string        @length(5)
  email:      string        @email @range(min: 5, max: 80)
  phones:     [PhoneNumber]

> }

Reviewer: ```suggestion
} @validate
```

Add this to align with the next example

The annotation examples above shows a validation scenario but annotations are not limited to this purpose. The developer has the freedom to extend the code generation tool to leverage annotations for their application's needs. In the `Customer` example, the developer could use these annotations to generate `validate` methods on each of the generated object types.

Directives are used to ensure that an annotation's arguments match an expected format. For constraining a numeric or length range, the directive for `@range` might look like this:

```apexlang
directive @range(min: u32, max: u32)
  on FIELD

>   require @valid on TYPE

Reviewer: ```suggestion
  require @validate on TYPE
```

```

> This directive specifies parameters but also where it can be applied. The syntax `on FIELD require @valid on TYPE` means "only allow this on fields where there is also an @valid annotation on its type". Multiple locations are supported and delimited with `|`:

Reviewer: ```suggestion
This directive specifies parameters but also where it can be applied. The syntax `on FIELD require @valid on TYPE` means "only allow this on fields where there is also an @validate annotation on its type". Multiple locations are supported and delimited with `|`:
```
